File: api/app/core/session_policy.py
"""Session lifetime policy.

R8 asks for a two-minute inactivity logout everywhere. Applied literally it
also logs out the vitals wall display, which exists to be watched and not
touched, and cuts off an elderly patient mid-dictation — the exact user the
voice feature is for. So the strict rule is kept where the threat actually is,
an unattended shared ward terminal, and the other classes are tiered.

See conflict C3 in the requirements triage.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from enum import StrEnum

from app.core.config import settings


class DeviceClass(StrEnum):
    SHARED_TERMINAL = "SHARED_TERMINAL"
    PERSONAL = "PERSONAL"
    MONITOR = "MONITOR"


#: Anything a client sends that is not recognised falls back to the strictest
#: tier, so a caller cannot widen its own timeout by inventing a device class.
DEFAULT_DEVICE_CLASS = DeviceClass.SHARED_TERMINAL

REFRESH_TOKEN_TTL_SECONDS = 60 * 60 * 24 * 7

#: How stale ``lastSeenAt`` may get before it is written again. Without this
#: every authenticated request would issue an UPDATE.
LAST_SEEN_WRITE_THROTTLE_SECONDS = 10


def coerce_device_class(value: str | None) -> DeviceClass:
    try:
        return DeviceClass(value or "")
    except ValueError:
        return DEFAULT_DEVICE_CLASS


def idle_timeout_seconds(device_class: str) -> int | None:
    """Seconds of inactivity allowed, or ``None`` when the class is exempt."""
    match coerce_device_class(device_class):
        case DeviceClass.MONITOR:
            # View-only wall display. Any action taken from it re-authenticates.
            return None
        case DeviceClass.PERSONAL:
            return max(settings.SESSION_IDLE_TIMEOUT_SECONDS, 15 * 60)
        case _:
            return settings.SESSION_IDLE_TIMEOUT_SECONDS


def access_token_ttl_seconds(device_class: str) -> int:
    """Kept at or below the idle window so a stolen token dies quickly."""
    idle = idle_timeout_seconds(device_class)
    return 15 * 60 if idle is None else min(idle, 15 * 60)


def absolute_timeout_seconds() -> int:
    return settings.SESSION_ABSOLUTE_TIMEOUT_SECONDS


@dataclass(frozen=True)
class IdleCheck:
    expired: bool
    #: Seconds remaining, or ``None`` when the class is exempt. Drives the
    #: client's countdown warning.
    remaining_seconds: int | None


def check_idle(device_class: str, last_seen_at: datetime, now: datetime | None = None) -> IdleCheck:
    idle = idle_timeout_seconds(device_class)
    if idle is None:
        return IdleCheck(expired=False, remaining_seconds=None)

    now = now or datetime.now(UTC)
    # Postgres `timestamp` columns come back naive; compare in the same frame.
    if last_seen_at.tzinfo is None:
        last_seen_at = last_seen_at.replace(tzinfo=UTC)
    if now.tzinfo is None:
        now = now.replace(tzinfo=UTC)

    elapsed = (now - last_seen_at).total_seconds()
    return IdleCheck(
        expired=elapsed >= idle,
        remaining_seconds=max(0, int(-(-(idle - elapsed) // 1))),
    )

File: server/src/modules/auth/session.policy.ts
import { env } from '../../config/env.js';

/**
 * Device classes and their idle timeouts.
 *
 * R8 asks for a two-minute inactivity logout everywhere. Applied literally it
 * also logs out the vitals wall display, which exists to be watched and not
 * touched, and cuts off a patient mid-dictation. So the strict rule is kept
 * exactly where the threat is — an unattended shared ward terminal — and the
 * other classes are tiered (see conflict C3 in the requirements triage).
 *
 * SHARED_TERMINAL is the default for any client that does not declare itself.
 */
export const DEVICE_CLASSES = ['SHARED_TERMINAL', 'PERSONAL', 'MONITOR'] as const;
export type DeviceClass = (typeof DEVICE_CLASSES)[number];

export const isDeviceClass = (value: unknown): value is DeviceClass =>
  typeof value === 'string' && (DEVICE_CLASSES as readonly string[]).includes(value);

/**
 * Idle seconds per class. SHARED_TERMINAL tracks the configured value so the
 * two-minute requirement stays a single knob in .env.
 * MONITOR is exempt from idle expiry: it is view-only, and any action taken
 * from it requires a fresh authentication.
 */
export const idleTimeoutSeconds = (deviceClass: string): number | null => {
  switch (deviceClass) {
    case 'MONITOR':
      return null;
    case 'PERSONAL':
      return Math.max(env.SESSION_IDLE_TIMEOUT_SECONDS, 15 * 60);
    case 'SHARED_TERMINAL':
    default:
      return env.SESSION_IDLE_TIMEOUT_SECONDS;
  }
};

/** Access-token lifetime. Kept at or below the idle window so a stolen token dies quickly. */
export const accessTokenTtlSeconds = (deviceClass: string): number => {
  const idle = idleTimeoutSeconds(deviceClass);
  if (idle === null) return 15 * 60;
  return Math.min(idle, 15 * 60);
};

export const REFRESH_TOKEN_TTL_SECONDS = 60 * 60 * 24 * 7;

export const absoluteTimeoutSeconds = (): number => env.SESSION_ABSOLUTE_TIMEOUT_SECONDS;

/**
 * How stale `lastSeenAt` may get before it is written again. Without this the
 * API would issue an UPDATE on every authenticated request; with it, activity
 * is still tracked far more finely than the shortest timeout.
 */
export const LAST_SEEN_WRITE_THROTTLE_MS = 10_000;

export interface IdleCheck {
  expired: boolean;
  /** Seconds of inactivity remaining, or null when the class is exempt. */
  remainingSeconds: number | null;
}

export const checkIdle = (deviceClass: string, lastSeenAt: Date, now = new Date()): IdleCheck => {
  const idle = idleTimeoutSeconds(deviceClass);
  if (idle === null) return { expired: false, remainingSeconds: null };
  const elapsedSeconds = (now.getTime() - lastSeenAt.getTime()) / 1000;
  return {
    expired: elapsedSeconds >= idle,
    remainingSeconds: Math.max(0, Math.ceil(idle - elapsedSeconds)),
  };
};
